### PR TITLE
Rule L019: Ignore comma placement violations if the adjacent code is templated

### LINKED ISSUE #2688: leading comma + dbt macro mistakenly produces "expected only leading" error
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

on a line like:

```
, 1 as int_col
, {{ mymacro('input') }} as special_column
, 'a' as str_col
```
I get an error: 
```
L:  2 | P:   5 | L019 | Found trailing comma. Expected only leading.  
```
It's only happening on columns with a macro reference and doesn't appear to correlate to the actual content of the macro. By that I mean, that even a macro that returns something as simple as an integer is generating the error. 

### Expected Behaviour

The error should not be raised since the expected leading comma exists

### Observed Behaviour

there appears to be a parsing error that isn't validating the leading comma. 

### How to reproduce

```
-- testfile.sql
select (
'a'
, {{ testmacro() }}
, 1 )

```
```
--testmacro.sql
{% macro testmacro() %}
1
{% endmacro %}
```

yields 
L:   3 | P:   1 | L019 | Found trailing comma. Expected only leading.

### Dialect

dbt + snowflake

### Version

0.10.1

### Configuration

[sqlfluff]
exclude_rules = L031, L033, L034, L023
select_clause_trailing_comma="forbid"
verbose = 0
nocolor = False
dialect = snowflake
templater = dbt
rules = None
recurse = 0
output_line_length = 120
runaway_limit = 10
ignore_templated_areas = True
encoding = autodetect
# Comma separated list of file extensions to lint.

# NB: This config will only apply in the root folder.
sql_file_exts = .sql,.sql.j2,.dml,.ddl

[sqlfluff:templater:dbt]
# project_dir = ./src
profiles_dir = ./src/profiles/
profile = snowflake-db

[sqlfluff:indentation]
indented_joins = True
indented_using_on = True
template_blocks_indent = True

[sqlfluff:templater]
unwrap_wrapped_queries = True

[sqlfluff:templater:jinja]
apply_dbt_builtins = True
load_macros_from_path= macros/

[sqlfluff:templater:jinja:macros]
# Macros provided as builtins for dbt projects
dbt_ref = {% macro ref(model_ref) %}{{model_ref}}{% endmacro %}
dbt_source = {% macro source(source_name, table) %}{{source_name}}_{{table}}{% endmacro %}
dbt_config = {% macro config() %}{% for k in kwargs %}{% endfor %}{% endmacro %}
dbt_var = {% macro var(variable, default='') %}item{% endmacro %}
dbt_is_incremental = {% macro is_incremental() %}True{% endmacro %}

# Some rules can be configured directly from the config common to other rules.
[sqlfluff:rules]
tab_space_size = 4
max_line_length = 100
indent_unit = space
comma_style = leading
allow_scalar = True
single_table_references = consistent
unquoted_identifiers_policy = all

# Some rules have their own specific config.
[sqlfluff:rules:L003]
indent_unit=space

[sqlfluff:rules:L007]  # Keywords
operator_new_lines = after

[sqlfluff:rules:L010]  # Keywords
capitalisation_policy = consistent

[sqlfluff:rules:L011]  # Aliasing
aliasing = explicit

[sqlfluff:rules:L012]  # Aliasing
aliasing = explicit

[sqlfluff:rules:L014]  # Unquoted identifiers
unquoted_identifiers_policy = all
extended_capitalisation_policy = upper

[sqlfluff:rules:L016]
ignore_comment_lines = False

[sqlfluff:rules:L026]
force_enable = False

[sqlfluff:rules:L028]
force_enable = False

[sqlfluff:rules:L029]  # Keyword identifiers
unquoted_identifiers_policy = aliases
quoted_identifiers_policy = none

[sqlfluff:rules:L030]  # Function names
capitalisation_policy = upper

[sqlfluff:rules:L038]
select_clause_trailing_comma = forbid

[sqlfluff:rules:L040]  # Null & Boolean Literals
capitalisation_policy = consistent

[sqlfluff:rules:L042]
# By default, allow subqueries in from clauses, but not join clauses.
forbid_subquery_in = join

[sqlfluff:rules:L047]  # Consistent syntax to count all rows
prefer_count_1 = False
prefer_count_0 = False

[sqlfluff:rules:L052]  # Semi-colon formatting approach.
semicolon_newline = False
require_final_semicolon = False

[sqlfluff:rules:L054]  # GROUP BY/ORDER BY column references.
group_by_and_order_by_style = consistent

ignore:

src/target/
/target
src/dbt_modules/
/dbt_modules
src/dbt_packages/
/dbt_packages
src/great_expectations/
raw_source_queries/
/great_expectations
dbt_packages/
.venv
macros/

### Are you willing to work on and submit a PR to address the issue?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

### LINKED ISSUE #2706: on dbt templater, dbt_utils.dateadd & dbt_utils.datediff don't seem to be parsed correctly
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

When linting my dbt repo for leading commas from the CLI, errors were found when lines contained dbt_utils.dateadd or dbt_utils.datediff as the main macro being applied. This doesn't seem to be a problem for other dbt macros. For example, these two lines both have leading commas, but the 2nd line flags a linting error for having a trailing comma:

```
        , {{ dbt_utils.surrogate_key(['log.time_stamp', 'log.connector_id', 'log.message_event']) }} AS row_id
        , {{ dbt_utils.datediff('log.time_stamp', 'current_timestamp', 'day') }} AS days_since
```

### Expected Behaviour

Both lines in the above should pass the linter for leading commas...

### Observed Behaviour

The 2nd line in the above failed the linter and was said to have a trailing comma

### How to reproduce

Trying to lint this sqlfile on a dbt-templater with the comma style set to leading (rule 019) produces a linting error on line 5:
```
WITH logs AS (
    SELECT
        1
        , {{ dbt_utils.surrogate_key(['1', '2']) }} AS row_id
        , {{ dbt_utils.datediff("'2021-01-01'", "'2021-05-04'", 'day') }} AS days_since
)

SELECT *
  FROM logs
```

### Dialect

bigquery

### Version

0.9.4 for sqlfluff, 1.0.0 for dbt

### Configuration

[sqlfluff]
verbose = 0
dialect = bigquery
templater = dbt
rules = L010, L019, L030, L040
ignore = parsing,templating

[sqlfluff:templater:dbt]
project_dir = .
#profiles_dir = ~/.dbt
profile = default
target = dev

[sqlfluff:rules]
tab_space_size = 4
max_line_length = 80
indent_unit = space

[sqlfluff:rules:L010]
# Keywords
capitalisation_policy = upper

[sqlfluff:rules:L019]
# Leading/Trailing Commas
comma_style = leading

[sqlfluff:rules:L030]
# Function names
capitalisation_policy = upper

[sqlfluff:rules:L040]
# Null & Boolean Literals
capitalisation_policy = upper

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by barrywhart**
### Brief summary of the change made
Fixes #2688, #2706 

### Pull Request checklist
- [x] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/dialects` (note YML files can be auto generated with `tox -e generate-fixture-yml`).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

### CODE REVIEW

File: src/sqlfluff/testing/rules.py [hunk 1]
```diff
@@ -94,17 +94,18 @@ def assert_rule_fail_in_sql(code, sql, configs=None, line_numbers=None):
...
     rendered = linter.render_string(sql, fname="<STR>", config=cfg, encoding="utf-8")
     parsed = linter.parse_rendered(rendered, recurse=True)
     if parsed.violations:
         pytest.fail(parsed.violations[0].desc() + "\n" + parsed.tree.stringify())
     print(f"Parsed:\n {parsed.tree.stringify()}")
-    lerrs, _, _, _ = r.crawl(
-        parsed.tree, [], dialect=cfg.get("dialect_obj"), templated_file=rendered[0]
-    )
+    lint_result = linter.lint_string(sql, config=cfg, fname="<STR>")
+    lerrs = lint_result.violations
```
Comment (barrywhart) [inline src/sqlfluff/testing/rules.py:108]: Fixing old tech debt that was breaking the new test: I changed this function to call the public `lint_string()` which runs the templater, parser, and linter in one step. The old code calls the rule **directly** which is hacky, and it was skipping the template-related filtering of lint results built into the linter, causing my new "pass" test to fail.

File: src/sqlfluff/rules/L019.py [hunk 2]
```diff
@@ -177,24 +177,32 @@ def _eval(self, context: RuleContext) -> Optional[LintResult]:
                                 )
                             ]
                         )
-                    return LintResult(
```
Comment (barrywhart) [inline src/sqlfluff/rules/L019.py:180]: It's best to review this with "Hide whitespace" enabled.

File: src/sqlfluff/core/rules/functional/segment_predicates.py [hunk 3]
```diff
@@ -168,7 +168,6 @@ def templated_slices(
         raise ValueError(
             'templated_slices: "templated_file" parameter is required.'
         )  # pragma: no cover
-    templated_slices = []
```
Comment (barrywhart) [inline src/sqlfluff/core/rules/functional/segment_predicates.py:171]: Unused variable

File: src/sqlfluff/core/rules/functional/segment_predicates.py [hunk 4]
```diff
@@ -179,9 +178,7 @@ def templated_slices(
     templated_slices = [
         slice_
         for slice_ in templated_file.sliced_file
-        if (
-            stop >= slice_.templated_slice.start and start < slice_.templated_slice.stop
-        )
+        if (stop > slice_.templated_slice.start and start < slice_.templated_slice.stop)
```
Comment (barrywhart) [inline src/sqlfluff/core/rules/functional/segment_predicates.py:181]: Fixed a boundary condition issue that was returning slices that it shouldn't

File: src/sqlfluff/core/rules/functional/templated_file_slices.py [hunk 5]
```diff
@@ -11,9 +11,9 @@ class TemplatedFileSlices(tuple):
     Provides useful operations on a sequence of slices to simplify rule creation.
     """
 
-    def __new__(cls, *raw_slices, templated_file=None):
+    def __new__(cls, *templated_slices, templated_file=None):
         """Override new operator."""
-        return super(TemplatedFileSlices, cls).__new__(cls, raw_slices)
+        return super(TemplatedFileSlices, cls).__new__(cls, templated_slices)
```
Comment (barrywhart) [inline src/sqlfluff/core/rules/functional/templated_file_slices.py:16]: Fix old copy-paste naming issue

File: src/sqlfluff/core/rules/functional/templated_file_slices.py [hunk 6]
```diff
@@ -35,3 +35,25 @@ def any(
...
+        """
+        start_index = self.index(start_slice) if start_slice else -1
+        stop_index = self.index(stop_slice) if stop_slice else len(self)
+        buff = []
+        for slice_ in self[start_index + 1 : stop_index]:
+            if loop_while is not None and not loop_while(slice_):
+                break
+            if select_if is None or select_if(slice_):
+                buff.append(slice_)
+        return TemplatedFileSlices(*buff, templated_file=self.templated_file)
```
Comment (barrywhart) [inline src/sqlfluff/core/rules/functional/templated_file_slices.py:59]: Added this method, equivalent to the `select()` methods on `Segments` and `RawFileSlices`. Ended up not using it, but thought it was worth keeping for future use.

File: src/sqlfluff/rules/L019.py [hunk 7]
```diff
@@ -134,7 +134,11 @@ def _eval(self, context: RuleContext) -> Optional[LintResult]:
             # A comma preceded by a new line == a leading comma
             if context.segment.is_type("comma"):
                 last_seg = self._last_code_seg(context.raw_stack)
-                if last_seg and last_seg.is_type("newline"):
+                if (
+                    last_seg
+                    and last_seg.is_type("newline")
+                    and not last_seg.is_templated
```
Comment (barrywhart) [inline src/sqlfluff/rules/L019.py:140]: This line is the actual fix: Ignore templated newlines when looking for rule violations.

**Review by tunetheweb — APPROVED** [review]

**Pull request merged**